some GCLowering cleanup and add support for FCAs

vtjnash opened the pull request:
In preparation for some future work, change some representations and cleanup some code.

vtjnash committed:
[GCLowering] Try to conceptually simplify numbering/lifting code.

Try to consolidate the code to share common sections between scalar and
vector inputs (in preparation for adding other aggregates).

Also fixes a lost root bug:
if a select or phi used a value from a select or phi, it would fail to root it.

Adds support for handling vectorized select condition in LiftSelect.

vchuravy commented on test/llvmpasses/gcroots.ll:
```diff
@@ -124,11 +125,12 @@ blabel:
+    %phi = phi i64 addrspace(12)* [ %adecayed, %alabel ], [ %bdecayed, %blabel ], [ %phi, %common ]
     call void @one_arg_decayed(i64 addrspace(12)* %phi)
-    ret void
+    br label %common
```
Why is this now a infinite for-loop?

vtjnash replied:
easiest way to make the test more complex without affecting anything that mattered

vchuravy commented on src/codegen.cpp:
```diff
@@ -5712,7 +5719,7 @@ static std::unique_ptr<Module> emit_function(
                     SP,                                 // Scope (current function will be fill in later)
                     jl_symbol_name(argname),            // Variable name
-                    ctx.has_sret + i + 1,               // Argument number (1-based)
+                    has_sret + i + 1,               // Argument number (1-based)
```
formatting of the comment

vchuravy commented on src/codegen.cpp:
```diff
@@ -5724,7 +5731,7 @@ static std::unique_ptr<Module> emit_function(
                     SP,                                 // Scope (current function will be fill in later)
                     std::string(jl_symbol_name(slot_symbol(ctx, ctx.vaSlot))) + "...",  // Variable name
-                    ctx.has_sret + nreq + 1,            // Argument number (1-based)
+                    has_sret + nreq + 1,            // Argument number (1-based)
```
formatting of the comment.

vtjnash commented on src/ccall.cpp:
```diff
@@ -673,6 +673,8 @@ static jl_cgval_t emit_cglobal(jl_codectx_t &ctx, jl_value_t **args, size_t narg
     }
     Type *lrt = julia_type_to_llvm(rt);
+    if (lrt == T_prjlvalue)
+        lrt = T_pjlvalue;
```
true, `assert(lrt != T_prjlvalue)` might be better

Keno commented:
Perhaps cherry-pick the `src: fix and sort makefile rules` onto master?

Keno commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -811,12 +804,13 @@ void LateLowerGCFrame::NoteUse(State &S, BBState &BBS, Value *V, BitVector &Uses
-        if (!isSpecialPtr(V->getType()) && !isSpecialPtrVec(V->getType()))
-            continue;
-        NoteUse(S, BBS, V, Uses);
+        if (isSpecialPtr(V->getType()) || isSpecialPtrVec(V->getType()) ||
```
I guess, this is the bugfix part of this commit? Add a test? I think `UpExposedUsesUnrooted` was left over from when we still used to optimize that. Also with test, I think this can be a separate PR, so we can get it in quicker.

Keno commented on test/llvmpasses/final-lower-gc.ll:
```diff
@@ -28,6 +28,7 @@ top:
 ; CHECK-NEXT: [[GCFRAME_SIZE_PTR2:%.*]] = bitcast %jl_value_t addrspace(10)** [[GCFRAME_SIZE_PTR]] to i64*
 ; CHECK-NEXT: store i64 4, i64* [[GCFRAME_SIZE_PTR2]], !tbaa !0
 ; CHECK-NEXT: [[GCFRAME_SLOT:%.*]] = getelementptr %jl_value_t**, %jl_value_t*** %ptls, i32 0
+; XXX: ordering may be inconsistent, causing the test to fail here on some platforms
```
Use `CHECK-DAG`?

vtjnash committed:
[GCLowering] consolidate vector and union handling

Permit handling any value with the same numbering scheme, not just Vector and Union (would also handle scalars, even though those typically go through a special case).

vtjnash commented:
Will merge ~tomorrow, if no comments.

StefanKarpinski commented:
Keno did leave some comments that appear to be unaddressed. If they're not relevant maybe comment why they're not?
